File: 1926-nearest-exit-from-entrance-in-maze/1926-nearest-exit-from-entrance-in-maze.py
class Solution:
    def nearestExit(self, maze: List[List[str]], entrance: List[int]) -> int:
        direction = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        q = deque()

        height, width = len(maze), len(maze[0])
        visited = set()
        q.append([entrance[0], entrance[1]])
        visited.add((entrance[0], entrance[1]))

        step = 0
        while q:
            for _ in range(len(q)):
                y, x = q.popleft()

                for dirY, dirX in direction:
                    locY = y+dirY
                    locX = x+dirX

                    if locY>=0 and locY<height and locX>=0 and locX<width and maze[locY][locX]=="." and (locY, locX) not in visited:
                        # return the exit once it is found
                        if locY==0 or locY==height-1 or locX==0 or locX==width-1:
                            return step+1
                        q.append([locY, locX])
                        # add the point to visited once it is found, 
                        # so that visited points won't be added to the queue again, like:
                        # . . .
                        # . . .
                        # . . .
                        visited.add((locY, locX))
            step += 1
        
        return -1

File: 1926-nearest-exit-from-entrance-in-maze/1926-nearest-exit-from-entrance-in-maze.cpp
class Solution {
public:
    int nearestExit(vector<vector<char>>& maze, vector<int>& entrance) {
        // tried to write it in cpp
        int height=maze.size(), width=maze[0].size();
        int dx[4] = {0, 0, 1, -1};
        int dy[4] = {1, -1, 0, 0};

        queue<pair<int, int>> q;
        q.push({entrance[0], entrance[1]});
        maze[entrance[0]][entrance[1]] = '+';

        int step = 0;
        while (!q.empty())
        {
            for (int size=q.size(); size>0; size--)
            {
                int y=q.front().first, x=q.front().second;
                q.pop();

                for (int i=0; i<4; i++)
                {
                    int loc_y = dy[i]+y;
                    int loc_x = dx[i]+x;
                    printf("%i, %i\n", loc_y, loc_x);
                    if (loc_y>=0 && loc_y<height && loc_x>=0 && loc_x<width && maze[loc_y][loc_x]=='.')
                    {
                        if (loc_y==0 || loc_y==height-1 || loc_x==0 || loc_x==width-1)
                            return step+1;
                        q.push({loc_y, loc_x});
                        maze[loc_y][loc_x] = '+';
                    }
                }
            }
            step++;
        }

        return -1;
    }
};